exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1903 @@
+	if err != nil {
+		return err
+	}
+	spotTickerPrice := ticker.Price{
+		AssetType:    asset.Spot,
+		Last:         response.Price,
+		LastUpdated:  response.Timestamp.Time(),
+		ExchangeName: ku.Name,
+		Pair:         pair,
+		Ask:          response.BestAsk,
+		Bid:          response.BestBid,
+		AskSize:      response.BestAskSize,
+		BidSize:      response.BestBidSize,
+		Volume:       response.Size,
+	}
+	assetEnabledPairs := ku.listOfAssetsCurrencyPairEnabledFor(pair)
+	if assetEnabledPairs[asset.Spot] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Spot) && ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+		ku.Websocket.DataHandler <- &spotTickerPrice
+	}
+	if assetEnabledPairs[asset.Margin] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Margin) && ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+		marginTickerPrice := spotTickerPrice
+		marginTickerPrice.AssetType = asset.Margin
+		ku.Websocket.DataHandler <- &marginTickerPrice
+	}
+	return nil
+}
+
+func (ku *Kucoin) processCandlesticks(respData []byte, instrument, intervalString string) error {
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	response := WsCandlestickData{}
+	err = json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	resp, err := response.getCandlestickData()
+	if err != nil {
+		return err
+	}
+	candlestickData := stream.KlineData{
+		Timestamp:  response.Time.Time(),
+		Pair:       pair,
+		AssetType:  asset.Spot,
+		Exchange:   ku.Name,
+		StartTime:  resp.Candles.StartTime,
+		Interval:   intervalString,
+		OpenPrice:  resp.Candles.OpenPrice,
+		ClosePrice: resp.Candles.ClosePrice,
+		HighPrice:  resp.Candles.HighPrice,
+		LowPrice:   resp.Candles.LowPrice,
+		Volume:     resp.Candles.TransactionVolume,
+	}
+	assetEnabledPairs := ku.listOfAssetsCurrencyPairEnabledFor(pair)
+	if assetEnabledPairs[asset.Spot] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Spot) && ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+		ku.Websocket.DataHandler <- candlestickData
+	}
+	if assetEnabledPairs[asset.Margin] && ku.AssetWebsocketSupport.IsAssetWebsocketSupported(asset.Margin) && ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+		candlestickData.AssetType = asset.Margin
+		ku.Websocket.DataHandler <- candlestickData
+	}
+	return nil
+}
+
+func (ku *Kucoin) processOrderbookWithDepth(respData []byte, instrument string) error {
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	response := WsLevel2Orderbook{}
+	err = json.Unmarshal(respData, &response)
+	if err != nil {
+		return err
+	}
+	asks := make([][3]string, len(response.Asks))
+	for x := range response.Asks {
+		asks[x][0] = response.Asks[x][0]
+		asks[x][1] = response.Asks[x][1]
+	}
+	bids := make([][3]string, len(response.Bids))
+	for x := range response.Bids {
+		bids[x][0] = response.Bids[x][0]
+		bids[x][1] = response.Bids[x][1]
+	}
```

Comment:
Just change respponse.Bids and Asks and allocate it as [][3]string as well. That way we don't need any double allocations, for loops and assignments here and this can be removed.